### PR TITLE
ISPN-9393 Support async methods with the Hot Rod near cache

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.jboss.org/browse/ISPN-9393

### CODE REVIEW

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/RemoteCacheImpl.java [hunk 1]
```diff
@@ -199,16 +199,21 @@ public boolean replaceWithVersion(K key, V newValue, long version, long lifespan
...
+   }
+
    @Override
    public MetadataValue<V> getWithMetadata(K key) {
-      return getWithMetadata(key, dataFormat);
+      return await(getWithMetadataAsync(key));
    }
 
-   private MetadataValue<V> getWithMetadata(K key, DataFormat dataFormat) {
+   protected CompletableFuture<MetadataValue<V>> getWithMetadataAsync0(K key, DataFormat dataFormat) {
```
Comment (rvansa) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/RemoteCacheImpl.java]: Other methods don't have the `0` suffix, why here?

File: documentation/src/main/asciidoc/user_guide/server_protocols_hotrod.adoc [hunk 2]
```diff
@@ -1163,6 +1163,9 @@ updated or removed server-side, the cached near cache entries are invalidated.
 If a key is requested after it's been invalidated, it'll have to be re-fetched
 from the server.
 
+WARNING: You should not use `maxIdle` expiration with near caches, as the
+behaviour is undefined.
```
Comment (rvansa) [inline documentation/src/main/asciidoc/user_guide/server_protocols_hotrod.adoc]: *Undefined* is convenient, but is there a reason to not say that accesses to the near cache will be ignored for the purposes of expiration?

Comment (tristantarrant) [reply documentation/src/main/asciidoc/user_guide/server_protocols_hotrod.adoc]: Yes, let me clear that. I also found a compilation failure.

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/RemoteCacheImpl.java (hunk 1 above)
Comment (tristantarrant) [reply client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/RemoteCacheImpl.java]: oops, a leftover

**Comment by tristantarrant** [comment]
Updated

**Comment by pruivo** [comment]
I guess `max-idle` can be supported if  `ClientCacheEntryExpired` is triggered and the near-cache service listens to it.

**Comment by tristantarrant** [comment]
@pruivo yes, but in order for maxIdle to work properly, every time the near cache does a local read, we would have to propagate the access time on the server as well, which sort of defeats the near cache

**Comment by rvansa** [comment]
@tristantarrant Not entirely, knowing maxIdle you could setup a timer on client to (deadline - 1 second) and issue an update command only at that time.
Not that I would think that it's worth the effort.

**Comment by tristantarrant** [comment]
Thanks but no thanks :)

**Comment by tristantarrant** [comment]
1 unrelated failure

**Comment by rvansa** [comment]
In that goes, then.

**Pull request merged**